### PR TITLE
Re-implement recognize, closes #1862

### LINKED ISSUE #1862: Router#recognize disappeared; no apparent way to provide an env to determine a route
As of [this commit](https://github.com/padrino/padrino-framework/commit/2d81e28cbd2b7172294acadcafe371734c8c4ea1#diff-c13c50f5426b23eb4d895160f07b3092L69), `Router#recognize` disappeared. At this point, there doesn't seem to be a way to pass a `Rack::MockRequest.env_for` hash and get a list of matching routes, and the padrino-rspec gem is broken as a result.

This means that tests can't determine whether something is routable (or if they can, then I'm no longer sure what the right way to do it is) by providing only a request env hash. It may be possible if you supply an entire request, but that is somewhat slower and would cause performance hits in tests if there are a lot of routes to check.

Is there a new recommended way to do this? For reference, the previously working RSpec matcher was along these lines:

```ruby
    matcher :be_routable do
      match do |verb_to_path_map|
        begin
          # compile Padrino's routing information
          Padrino.application.call(Rack::MockRequest.env_for("/"))
        end
        path = verb_to_path_map.values.first
        method = verb_to_path_map.keys.first.to_s.upcase
        @routed_to = Padrino.mounted_apps.map(&:app_obj).
          map{|a| a.router.recognize(Rack::MockRequest.env_for(path, :method => method))}.first
        @routed_to.first
      end

      failure_message_for_should_not do |path|
        "expected #{path.inspect} not to be routable, but it routes to #{@routed_to.inspect}"
      end
    end
```

@dariocravero suggested doing a workaround of duck-typing an object that looks like what Padrino wants, so one can do something like this:

```ruby
  class RouterMockRequest
    attr_accessor :path_info, :request_method, :params

    def initialize(path_info: '/', request_method: :get, params: {})
      @params         = params
      @path_info      = path_info
      @request_method = request_method
    end
  end
```

and then one can do

```ruby
    @routed_to = Padrino.mounted_apps.map(&:app_obj).
      map{|a| a.router.engine.find_by_request(RouterMockRequest.new(path_info: path, request_method: method)) {}}.first
```

which will work to get tests passing again. However, this couples all of my routing tests to internal details of Padrino's implementation, which isn't preferable.

### PR DESCRIPTION
**Pull request by namusyaka**
ref #1862
Implemented PathRouter#recognize to recognize routes by using request or env without block.
I think this can resolve the problem.
@fj @dariocravero What do you think?

And, I found a little bit problem about compatibility with old routing, so I also fixed it.

### CODE REVIEW

File: padrino-core/lib/padrino-core/path_router/compiler.rb [hunk 1]
```diff
@@ -34,12 +34,23 @@ def compiled?
...
+
       ##
-      # Finds routes by using request.
+      # Calls routes by using request.
       #
-      def find_by_request(request)
+      def call_by_request(request)
         rotation do |offset|
-          pattern  = request.path_info.encode(Encoding.default_external)
+          pattern  = encode_default_external(request.path_info)
```
Comment (dariocravero) [inline padrino-core/lib/padrino-core/path_router/compiler.rb]: Couldn't this be outside `rotation`?

File: padrino-core/lib/padrino-core/path_router/compiler.rb [hunk 2]
```diff
@@ -34,12 +34,23 @@ def compiled?
...
       end
   
+
+      ##
+      # Finds routes by using request or env.
+      #
+      def find_by(request_or_env)
+        request = request_or_env.is_a?(Hash) ? Sinatra::Request.new(request_or_env) : request_or_env
+        pattern = encode_default_external(request.path_info)
+        verb    = request.request_method.downcase.to_sym
```
Comment (dariocravero) [inline padrino-core/lib/padrino-core/path_router/compiler.rb]: Should we normalise this bit [here too](https://github.com/padrino/padrino-framework/pull/1863/files#diff-7de007036e25d75ce0d909f152958551R56) and either extract `verb` everywhere or use `request.request_method.downcase.to_sym`?

File: padrino-core/lib/padrino-core/path_router/compiler.rb [hunk 3]
```diff
@@ -34,12 +34,23 @@ def compiled?
...
+      # Calls routes by using request.
       #
-      def find_by_request(request)
+      def call_by_request(request)
         rotation do |offset|
-          pattern  = request.path_info.encode(Encoding.default_external)
+          pattern  = encode_default_external(request.path_info)
           if route = match?(offset, pattern)
             params = route.params_for(pattern, request.params)
             yield(route, params) if route.verb == request.request_method.downcase.to_sym
```
Comment (dariocravero) [inline padrino-core/lib/padrino-core/path_router/compiler.rb]: After this, `route` keeps on being returned regardless it matching the verb or not. Maybe I'm missing something but if we only run its block when the verb matches, why do we need to return it anyway?

**Comment by dariocravero** [comment]
Thanks for putting this together @namusyaka :). :clap: Apart from the comments above, I think it's good to go :)

File: padrino-core/lib/padrino-core/path_router/compiler.rb (hunk 2 above)
Comment (namusyaka) [inline padrino-core/lib/padrino-core/path_router/compiler.rb]: Right. We don't have to symbolize request methods.

File: padrino-core/lib/padrino-core/path_router/compiler.rb (hunk 1 above)
Comment (namusyaka) [inline padrino-core/lib/padrino-core/path_router/compiler.rb]: Nope, it is on purpose.
The request instance must be able to modify to ensure a compatibility with sinatra.
See https://github.com/padrino/padrino-framework/blob/master/padrino-core/test/test_filters.rb#L390

File: padrino-core/lib/padrino-core/path_router/compiler.rb (hunk 3 above)
Comment (namusyaka) [inline padrino-core/lib/padrino-core/path_router/compiler.rb]: In order to implement 405 response, all matched routes should be returned even if doesn't match with verb.
Of course, we could avoid returning matched routes but it isn't intuitive.
Do I understand correctly what you say?

**Comment by fj** [comment]
Hi @namusyaka -- this fixes my tests in my application that I mentioned #1862, so :+1: from me. Thanks for your hard work!

**Pull request merged**